### DISCUSSION TITLE
JSON datasource with separate time series

### DISCUSSION THREAD

**Discussion by stevisco** [grafana/grafana-infinity-datasource#275] [Q&A]
Hi, I have a JSON datasource (a REST API) which returns multiple timeseries as separate objects. See the example below. Is it possible to use it with this plugin ? 

`{
  "series1": [
    { "time": "2022-01-01T00:00:00.000Z", "value": 22},
    { "time": "2022-01-02T00:00:00.000Z", "value": 23 }
  ]
 },
{
  "series2": [
    { "time": "2022-01-01T00:00:00.000Z", "value": 220},
    { "time": "2022-01-02T00:00:00.000Z", "value": 230 }
  ]
 }
`

**Comment by yesoreyeram** [comment]
Sorry for the delay. Still if you need help, I am able to get this done with UQL

```json
{
  "series1":[
         {"time":"2022-01-01T00:00:00.000Z","value":22},
         {"time":"2022-01-02T00:00:00.000Z","value":23}
    ],
  "series2":[
         {"time":"2022-01-01T00:00:00.000Z","value":220},
         {"time":"2022-01-02T00:00:00.000Z","value":230}
    ]
}
```

and the UQL

```kusto
parse-json
| project kv()
| mv-expand "value"
| project "time"=todatetime("value.time"), "value"="value.value", "series"="key"
```

you can see the result [here](http://localhost:3000/explore?orgId=1&left=%7B%22datasource%22:%22Infinity%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22type%22:%22uql%22,%22source%22:%22inline%22,%22format%22:%22timeseries%22,%22url%22:%22https:%2F%2Fjsonplaceholder.typicode.com%2Fusers%22,%22url_options%22:%7B%22method%22:%22GET%22,%22data%22:%22%22%7D,%22root_selector%22:%22%22,%22columns%22:%5B%5D,%22filters%22:%5B%5D,%22global_query_id%22:%22%22,%22data%22:%22%7B%5C%22series1%5C%22:%5B%7B%5C%22time%5C%22:%5C%222022-01-01T00:00:00.000Z%5C%22,%5C%22value%5C%22:22%7D,%7B%5C%22time%5C%22:%5C%222022-01-02T00:00:00.000Z%5C%22,%5C%22value%5C%22:23%7D%5D,%5C%22series2%5C%22:%5B%7B%5C%22time%5C%22:%5C%222022-01-01T00:00:00.000Z%5C%22,%5C%22value%5C%22:220%7D,%7B%5C%22time%5C%22:%5C%222022-01-02T00:00:00.000Z%5C%22,%5C%22value%5C%22:230%7D%5D%7D%22,%22uql%22:%22parse-json%5Cn%7C%20project%20kv()%5Cn%7C%20mv-expand%20%5C%22value%5C%22%5Cn%7C%20project%20%5C%22time%5C%22%3Dtodatetime(%5C%22value.time%5C%22),%20%5C%22value%5C%22%3D%5C%22value.value%5C%22,%20%5C%22series%5C%22%3D%5C%22key%5C%22%22%7D%5D,%22range%22:%7B%22from%22:%221640995200000%22,%22to%22:%221641081600000%22%7D%7D)

<img width="1791" alt="image" src="https://user-images.githubusercontent.com/153843/174670760-0dbb4f5c-f303-4a4b-a1a3-484b010a9893.png">

**Reply by svet-b** [reply]
I have a similar case, just with an additional level in the JSON, e.g.:
```
{
  "series1": {
    "data": [
         {"time":"2022-01-01T00:00:00.000Z","value":22},
         {"time":"2022-01-02T00:00:00.000Z","value":23}
    ]
  },
  "series2": {
    "data": [
         {"time":"2022-01-01T00:00:00.000Z","value":220},
         {"time":"2022-01-02T00:00:00.000Z","value":230}
    ]
  }
}
```
(omitting additional fields)

Is there a way to do this in UQL? I tried various things, such as adding `| scope "data"` either before or after the `| mv-expand "value"` line, or prepending `data.` to the selectors in `| project`, but didn't get anywhere useful. Any tips?

By the way, is there a way to debug UQL statements - e.g. to simply display the "draft" output after a set of directives, as I build up the query? The process of composing the query (and hoping that it outputs something that Grafana can show me, e.g. in a table) feels like a lot of stabbing around in the dark, but maybe I'm missing something? Maybe I didn't look hard enough within the query inspector?

**Reply by svet-b** [reply]
FWIW the JSONata expression
```
$each($, function($v, $s) {[$v.data.{'series': $s, 'time':$.time, 'value': $.value}]}).*
```
produces appropriate output in the case I described, namely breaking out all of the individual data points into an array of objects:
```
[
  {
    "series": "series1",
    "time": "2022-01-01T00:00:00.000Z",
    "value": 22
  },
  {
    "series": "series1",
    "time": "2022-01-02T00:00:00.000Z",
    "value": 23
  },
  {
    "series": "series2",
    "time": "2022-01-01T00:00:00.000Z",
    "value": 220
  },
  {
    "series": "series2",
    "time": "2022-01-02T00:00:00.000Z",
    "value": 230
  }
]

**Reply by yesoreyeram** [reply]
I didn't investigate much.. but I would say JSONata should work.  ( trick is to choose Timeseries as format )

```kusto
parse-json
| jsonata "$each($, function($v, $s) {[$v.data.{'series': $s, 'time':$.time, 'value': $.value}]}).*"
```

<img width="1635" alt="image" src="https://user-images.githubusercontent.com/153843/207327193-8664128c-ce4e-48ed-984a-5ed3930f21b5.png">

**Reply by svet-b** [reply]
Yes sorry I wasn't explicit in my last reply - the JSONata expression did work perfectly, exactly as you describe in your example 👍

**Comment by maddsua** [comment]
Hey, @yesoreyeram!
Say I have full control over the data that gets sent to infinity, what would be the format of the data to achieve the same result?
My goal here is to avoid doing unnecessary transforms when the upstream can just output the data in a needed format

**Reply by yesoreyeram** [reply]
Ideally the formats specified in the contract 
https://grafana.com/developers/dataplane/

**Reply by yosiasz** [reply]
why does it add the value string to the metrics?

![image](https://github.com/user-attachments/assets/a47b5fd1-601e-4d8c-8a50-becc4f65379d)

**Reply by yosiasz** [reply]
no issue when querying it in sql server

![image](https://github.com/user-attachments/assets/45f712a0-5195-4c20-91cb-8fe18e5ced22)
